components/tikv_util/src/memory.rs

```diff
@@ -205,55 +216,31 @@ impl MemoryQuota {
     }
 
     pub fn alloc_force(&self, bytes: usize) {
-        let mut in_use_bytes = self.in_use.load(Ordering::Relaxed);
-        loop {
-            let new_in_use_bytes = in_use_bytes + bytes;
-            match self.in_use.compare_exchange_weak(
-                in_use_bytes,
-                new_in_use_bytes,
-                Ordering::Relaxed,
-                Ordering::Relaxed,
-            ) {
-                Ok(_) => return,
-                Err(current) => in_use_bytes = current,
-            }
-        }
+        self.in_use.fetch_add(bytes as isize, Ordering::Relaxed);
     }
 
     pub fn alloc(&self, bytes: usize) -> Result<(), MemoryQuotaExceeded> {
-        let capacity = self.capacity.load(Ordering::Relaxed);
-        let mut in_use_bytes = self.in_use.load(Ordering::Relaxed);
-        loop {
-            if in_use_bytes + bytes > capacity {
-                return Err(MemoryQuotaExceeded);
-            }
-            let new_in_use_bytes = in_use_bytes + bytes;
-            match self.in_use.compare_exchange_weak(
-                in_use_bytes,
-                new_in_use_bytes,
-                Ordering::Relaxed,
-                Ordering::Relaxed,
-            ) {
-                Ok(_) => return Ok(()),
-                Err(current) => in_use_bytes = current,
-            }
+        let capacity = self.capacity.load(Ordering::Relaxed) as isize;
+        let in_use_bytes = self.in_use.load(Ordering::Relaxed);
+        if bytes > capacity.saturating_sub(in_use_bytes) as usize {
+            return Err(MemoryQuotaExceeded);
+        }
+        let bytes = bytes as isize;
+        let new_in_use_bytes = self.in_use.fetch_add(bytes, Ordering::Relaxed);
+        if bytes > capacity - new_in_use_bytes {
+            self.in_use.fetch_sub(bytes, Ordering::Relaxed);
+            return Err(MemoryQuotaExceeded);
         }
+        Ok(())
     }
 
     pub fn free(&self, bytes: usize) {
-        let mut in_use_bytes = self.in_use.load(Ordering::Relaxed);
-        loop {
-            // Saturating at the numeric bounds instead of overflowing.
-            let new_in_use_bytes = in_use_bytes - std::cmp::min(bytes, in_use_bytes);
-            match self.in_use.compare_exchange_weak(
-                in_use_bytes,
-                new_in_use_bytes,
-                Ordering::Relaxed,
-                Ordering::Relaxed,
-            ) {
-                Ok(_) => return,
-                Err(current) => in_use_bytes = current,
-            }
+        let bytes = bytes as isize;
```

Comment:
What if the bytes is larger than `isize::MAX`?

Reply:
Done. In order to handle overflow, I added a hard limit for a single alloc call.